### PR TITLE
implement use case 2.5 on fMRI data

### PR DESCRIPTION
**Pull request by Marque-CerCo**

### CODE REVIEW

**Comment by marcelzwiers** [comment]
Hi Adrian,

You added your code in a `wp15` subdirectory, which is confusing because that directory also exists as a parent directory. Could you please rename it to something more meaningful (like `source` or `code` or something)? Also, I was wondering why it is needed to change the permissions of the files? It is not needed for the other open-neuro datasets... Did you solve the problem of needing to edit SPM? In that context, perhaps this is also of help:

https://www.fil.ion.ucl.ac.uk/spm/docs/wikibooks/Headless/

**Comment by Marque-CerCo** [comment]
Hello,
I renamed wp15 to source.
 I solved the problems with SPM and considered your comments by modifying the Matlab command line. I'm now testing the changes and checking whether I need to change the permissions. If so, I'll get back to you to explain why.
Sincerely

**Comment by Marque-CerCo** [comment]
When I download the database with Datalad, the patient files appear as *.nii.gz files. These are symbolic link files. The unzipDatabase.py file allows you to :
- retrieve files via a symbolic link 
- decompress them
The changePermissions.py file is used to modify permissions because once the files are unzipped, I can only read the unzipped files (in memory).
Sincerely

**Comment by marcelzwiers** [comment]
I don't think this is how you should work with datalad data. Datalab indeed locks the data and present a symbolic link data skeleton to the user. This is intentional, to allow Datalab to do its data versioning. You should not interfere with that by changing permissions etc., but instead use the DataLad mechanism to unlock the data

http://docs.datalad.org/en/stable/generated/man/datalad-unlock.html

**Comment by Marque-CerCo** [comment]
I appreciate your help. 
I followed your recommendations and tested the _**datalad unlock_** command line. I tested this command line on the database. I made some modifications by deleting the Python files and modifying the _**datalad**_ command line.
Sincerely

**Comment by Marque-CerCo** [comment]
Can a git clone command line be provided for the source directory? Do I need to reorganize the Matlab files?

**Comment by schoffelen** [comment]
In its current state, this branch does not contain any functional code that can be tested off-site. @Marque-CerCo will you let us (i.e. @CPernet @marcelzwiers etc) know once the code is available? It's fine to merge everything in the master branch of the wp15 repo as far as I am concerned, rather than keeping it in this PR branch

**Comment by Marque-CerCo** [comment]
Hello,
While making the changes mentioned in Friday's discussion, I realized the database's limitations. So, I looked for a new database on OpenNeuro and adapted my code to it. I also took the opportunity to simplify and clean up my code (improving computation time). 
What should I describe in the "pipeline" and "output" sections? Can a scheme be created (with github) to explain output directory architecture? 
Sincerely

**Comment by schoffelen** [comment]
Hi @Marque-CerCo can you please confirm that the openneuro dataset that should be used is ds004934 (as suggested by the description and the reference to the paper in the README.md), rather than the ds003020 (as suggested by the 'git clone' instruction in a previous version of the README.md). I have by now pushed some changes into the README.md file in the open pull request. Please pull those changes locally before pushing any other changes to the code.

**Comment by Marque-CerCo** [comment]
The openneuro dataset is ds004934.

**Comment by schoffelen** [comment]
Although it could be still a bit premature, I tried to follow the instructions in the current version of the README.md, but I got stuck. I managed to download the data, but the ```git clone source``` does not work. This probably should be a command to clone the wp15 repository, which contains usecase-2.5/source. Also, when I try to run WorkPackageCerCo.m file, I quickly get into trouble because there are some specific assumptions on the location from where the script is run, and on the directory organisation w.r.t. where the input data is relative to the code. I committed some suggested changes to the code to make it a bit more robust, but please feel free to organize it differently as you see fit. At least it is important to remove any direct pointers to specific users' home-directory, as was the case previously. Next, with the path issues resolved I realised that I could not quickly test a run of the pipeline for a single participant, because the downloaded files are still gzipped. I presume that the intended python code for unzipping is still work in progress, but you may consider to do the unzipping in matlab, using the gunzip function. Something like the following would do the trick:

```
cd(input_path)
subs = dir('sub-*/*/*.gz');
for k = 1:numel(subs)
  gunzip(fullfile(subs(k).folder, subs(k).name);
end
```

**Comment by Marque-CerCo** [comment]
Hi @schoffelen,
I've accepted your comments on the code's robustness. For the "git clone source" command, I asked earlier if providing a link to download the source directory was possible. I want to offer future users a link so that all Matlab files are available via git clone. 
For the data, I use datalad via the following commands:
cd input
datalad get sub-*
datalad unlock sub-*
Can you see these command lines? 
I'll test these lines to see if I get an error message.

**Comment by schoffelen** [comment]
Hi @Marque-CerCo datalad works fine for me. The instructions look good, although I don't understand (yet) why the unlocking is needed). I executed those commands, and got the data just fine. Yet, I needed to unzip the *.nii.gz files  before I could proceed. Now, with respect to cloning the repo, for the usecase one should be able to do ```git clone https://github.com/SIESTA-eu/wp15```. This will pull in the wp15 repository, which contains a folder usecase-2.5/source/ where the relevant matlab code for this usecase is stored. That should be sufficient, shouldn't it?

**Comment by schoffelen** [comment]
> Hello, While making the changes mentioned in Friday's discussion, I realized the database's limitations. So, I looked for a new database on OpenNeuro and adapted my code to it. I also took the opportunity to simplify and clean up my code (improving computation time). What should I describe in the "pipeline" and "output" sections? Can a scheme be created (with github) to explain output directory architecture? Sincerely

Hi @Marque-CerCo you can use the ```tree``` command on the linux command line to sketch the organization of a directory. It comes with a bunch of options, if you type ```man tree``` you will see what can be done.

For example, my working dir for siesta looks like:

```
jansch@mentat006:~/project_siesta/jansch
1029 $ tree -L 1
.
|-- data
|-- envs
|-- input
|-- output
|-- sandbox.py
|-- siesta
|-- spm12
|-- venv
`-- wp15
```

**Comment by Marque-CerCo** [comment]
Hi @schoffelen, 
With detailed, *.nii.gz files are associated with symbolic links. The unlock command retrieves files via symbolic links. I made a mistake when it came to unzip files. I have attached an old version of "workPackageCerCo.m". 
So, I added a matlab code to unzip *.nii.gz files.

**Comment by Marque-CerCo** [comment]
> > Hello, While making the changes mentioned in Friday's discussion, I realized the database's limitations. So, I looked for a new database on OpenNeuro and adapted my code to it. I also took the opportunity to simplify and clean up my code (improving computation time). What should I describe in the "pipeline" and "output" sections? Can a scheme be created (with github) to explain output directory architecture? Sincerely
> 
> Hi @Marque-CerCo you can use the `tree` command on the linux command line to sketch the organization of a directory. It comes with a bunch of options, if you type `man tree` you will see what can be done. I am not sure whether the command is installed on your system by default.
> 
> For example, my working dir for siesta looks like:
> 
> ```
> jansch@mentat006:~/project_siesta/jansch
> 1029 $ tree -L 1
> .
> |-- data
> |-- envs
> |-- input
> |-- output
> |-- sandbox.py
> |-- siesta
> |-- spm12
> |-- venv
> `-- wp15
> ```
> 
> For the purpose of fooling around with the PR that you have been working on, the relevant directories shown above, are input/output/siesta (which contains the virtual environment's specification+data)/spm12/wp15.
> 
> I took the liberty of downloading spm12 directly from their github (for now, let's see where I get shipwrecked), rather than the one that you suggest in the README.md. The most important reason for this is that somehow it has become unclear in the CerCo version of spm12 what has changed in detail, relative to the original code. Probably, this information was lost along the way because the files were added 'by hand', using some kind of GUI, without a revision history.

Thank you for your reply. I will look the 'tree command'.

File: usecase-2.5/source/workPackageCerCo.m [hunk 1]
```diff
@@ -29,6 +29,16 @@
 addpath(path_matlabbatch);
 addpath(path_workpackage);
 
+% unzip nifti files
+cd(path_input)
+subs = dir('sub-*/*/*.gz');
+for k = 1:numel(subs)
+    path_file = fullfile(subs(k).folder, subs(k).name);
+    disp(path_file);
+    gunzip(fullfile(subs(k).folder, subs(k).name));
```
[the remark this answers is not in the document]
Comment (Marque-CerCo) [reply usecase-2.5/source/workPackageCerCo.m:38]: Hello,
I added a line to delete *.nii.gz files. I also changed _git clone source_ to _git clone https://github.com/SIESTA-eu/wp15_. Then, I adapter the matlab command line.

[the remark this answers is not in the document]
Comment (schoffelen) [reply usecase-2.5/source/workPackageCerCo.m:38]: OK thanks.

**Comment by schoffelen** [comment]
Hi @Marque-CerCo I made a few (commented) commits to the codebase that intend to make testing a bit easier from our end. Also, I suggested a change which probably is a fix: in the earlier version, the onePatient function did the preprocessing on just a single subject, but the parameter + contrast estimation in a loop across all subjects. Please correct me if I am wrong, but this is not desired behavior, right? Also, for the purpose of efficiency, I suggest to do the check for the presence of *.gz files closer to the actual computations (in spm_preprocess), rather than for the whole database. Moreover, stylistically, a lot of the code evaluates the contents of a directory by a call to the function ```dir``` followed by some additional logical evaluations w.r.t. the file name and/or the file's extension. I think that this can/should be done in a single loop (e.g. ```dir(fullfile(somepathname, 'sub*nii'))``` to select all *.nii file whose name starts with sub). This - as a by-effect - avoids a situation that I ended up in recently, when I tried to run spm_preprocess for the second time on a folder which already contained partially processed results, causing a recursion in that the code tried to realign/coregister etc. already realigned/coregistered files (because initially it performed the preprocessing steps on all nii files in a given directory). This brings me to a conceptually relevant point: should the preprocessed results be stored in the 'input' or in the 'output' directory. My hacky way of testing the pipeline caused me troubles given the previous state of code and data organization, and one could argue that this is too bad for me (because in a putative production stage the pipeline is ran only once), although in the grand scheme of things it may make sense to think a bit more broadly and adopt programming/data organisation strategies that are more widely applicable without too much hassle.
I made an initial attempt to allow the pipeline to be run on a single (or small subset of) subject(s), so that the pipeline can be more efficiently run multiple times. I realise that in the current dataset there is some room for additional speed up, i.e. we could use similar logic to also allow for the computation for only a subset of the runs, and/or tasks. 
Also, am I right in concluding that in the present state of the pipeline there is not yet any code that is computing second-level results, i.e. results that are to be the actual 'output' of the pipeline?

**Comment by schoffelen** [comment]
OK, I had another pass at this, and it is getting into shape a bit better.
I have adjusted the code a bit, so that the events.tsv files can be read directly.
Also, there is now the possbility to run a subset of subjects, tasks, runs

What I don't like very much, is that I now ran into - I think - the motivation for using an adjusted version of SPM. The motivation seems to be that - if an SPM.mat file already exists in the specified output location - a user interaction is required from the normal SPM. It looks as if the adjusted version has a call to spm_input commented out. I think that this should not be done. We should use normal SPM, and then deal with the possibility that the SPM.mat already exists in our own code.

**Comment by Marque-CerCo** [comment]
Hello @schoffelen, 
I am reading the changes made to the code. I saw that you adapted the code. To answer one of your questions, I store the intermediate files related to the spm_process function in the input directory. I delete these files at the end with the cleanDatabase function. I can delete these files once they have been used. The results are in the output directory with the SPM.mat file and the beta_*.nii files.
Do you prefer that I use the standard version of spm and not the version I propose?

**Comment by Marque-CerCo** [comment]
I can also save the results of the spm_preprocess function in the output directory.

**Comment by schoffelen** [comment]
> I can also save the results of the spm_preprocess function in the output directory.

I think that with more careful checking of the filenames (specifically in the realign step) it is also fine to keep it as is.

**Comment by Marque-CerCo** [comment]
> > I can also save the results of the spm_preprocess function in the output directory.
> 
> I think that with more careful checking of the filenames (specifically in the realign step) it is also fine to keep it as is.

So, I am going to download the code you have modified. Then, I will save the files from the different steps to the corresponding location in the output directory.

**Comment by schoffelen** [comment]
>  Do you prefer that I use the standard version of spm and not the version I propose?

Yes. I just realized that 'your' version is just plain SPM with one (or a few) lines commented out. This is not good. We should be able to use normal SPM, and deal with the cause of the interactive questions (as far as I observed this is caused by some files already existing, but perhaps some other causes too) in our code, not by commenting out the SPM code

**Comment by schoffelen** [comment]
> > > I can also save the results of the spm_preprocess function in the output directory.
> > 
> > 
> > I think that with more careful checking of the filenames (specifically in the realign step) it is also fine to keep it as is.
> 
> So, I am going to download the code you have modified. Then, I will save the files from the different steps to the corresponding location in the output directory.

I am not sure what you mean with 'save the files from the different steps...' We should not start moving around 'by hand' generated output data. We should be testing and re-testing the pipeline until it works in a satisfactory way, independently from where it will be executed. I built in the possibility for running subsets of subjects etc. to have a faster test cycle in general (specifically on my end, because by nature I am a bit impatient, and don't want to wait for a full iteration of all subjects, tasks, runs)

Also, we should get some second-level test results, which are the intended outputs to be specified on the whitelist.

**Comment by Marque-CerCo** [comment]
Hi @schoffelen ,
Here are the changes I've made:
- input and output directories,
- the code works with the standard version of spm (not mine)
I have a problem with the contrast function, which won't let me proceed to the rest of the code. The code executes but doesn't generate the desired files. Once everything is working, I'll propose this code.

**Comment by schoffelen** [comment]
Hi @Marque-CerCo what do you mean with 'the code'? On my end it still does not work (i.e. the most recent code pulled from the PR -> I created a remote on my end, that directly pulls from the patch-1 branch), even if I work in an interactive session, where I can tell SPM to proceed (in case of an already existing SPM.mat in a subject output folder, because of my multiple attempts).

**Comment by schoffelen** [comment]
What is the plan de campagne to push this forward? I might have missed this, but I did not see any updates following up on my written updates in the meeting notes/github. I would like to avoid the situation that we are waiting for each other to bring the usecase 2.5 to a basic working completion. Of course I could try and hack my way through, but I’d rather not break anything. Also, since I am unaware of the intended second level outcomes of this pipeline, I’d rather not phantasize something together.

**Comment by schoffelen** [comment]
Hi @Marque-CerCo I started a fresh environment, and following the README in its currently proposed state I tried to execute the workPackageCerCo.m function (calling it in an interactive matlab session, without input arguments). 

I had to make a small fix in the patientsDatabase function in order to make it past line 10, but then I got an error, caused by the fact that the onePatient function is called (in line 12) with 4 input arguments, but the current version of onePatient in the PR is back to its original dysfunctional state (not only with just 3 input arguments, but also with the - I think flawed - logic of computing first level results of all subjects (while preprocessing only a single subject), followed by a secondlevel step.
I would appreciate it if we could move the code forward so that :
- the first level computations (i.e. preprocessing and first level model fit and so on) can be optionally done for a subset of participants, a subset of runs and a subset of tasks.
- the second level computations are done in a separate function, i.e. the function secondlevel is NOT to be called in the onePatient function.

I can make the desired changes, but I'd rather make them just a single time, alternatively you stay in the driver seat, and make the necessary changes before I give it a new try. Please let me know how you wish to proceed.

**Comment by Marque-CerCo** [comment]
Hi @schoffelen, 
You can modify the files because you have a specific idea (for the first-level computations). I fully understand the idea of adapting the code to a subset of participants. But, what do you mean by a subset of runs and a subset of tasks?

**Comment by schoffelen** [comment]
Well, I got the impression from the data that there are a few tasks in the dataset, with a few runs per task. To speed up prototyping, it would be nice to be able to run the pipeline on not only a subset of participants but also on a subste of the tasks.

**Comment by Marque-CerCo** [comment]
For example, you want to limit the pipeline to patient 1 for the Motion task by applying the spm_preprocess and first-level functions. Is it correct ?

**Comment by schoffelen** [comment]
hi @Marque-CerCo @CPernet @marcelzwiers @robertoostenveld I have made extensive changes to the code, rewrote the firstLevel/secondLevel functions (because I did not understand the heuristics used in the original ones for building the first/second level models, got in trouble w.r.t. out-of-range while loops etc among others), and did an initial test run on our HPC. When I run: ```workPackageCerCo({'sub-SAXNES2s001';'sub-SAXNES2s002';'sub-SAXNES2s003'}); ``` I now get some output data in the ```output/groupresults``` folder. Some of these files are most likely 'relevant' group level files to be whitelisted. 

The next steps for this use case now are:
1) (optional): further clean up the code, it's now a bit messy, but if you ask me, I wouldn't touch it too extensively, at the risk of breaking it again, in which case I am not going to fix it anymore
2) (imperative): pull the code from this PR - although I would suggest to merge this PR first, and then pull the main branch - and test whether the current state of the code runs somewhere else as well.
3) complete the README.md with the necessary information
4) build the pipeline + matlab into a container so that the tech folks can run it more easily

**Pull request merged**